tests/utils.py

```diff
@@ -1,40 +1,182 @@
 from __future__ import annotations
 
 import math
+from abc import abstractmethod
```

Comment:
Most of the changes can be moved into a separate pull request if the approach as a whole is satisfactory.